buildpacks/scala/src/main.rs

```diff
@@ -0,0 +1,415 @@
+
+        BuildResultBuilder::new().build()
+    }
+
+    fn on_error(&self, error: Error<Self::Error>) {
+        on_buildpack_error(log_user_errors, error);
+    }
+}
+
+buildpack_main!(ScalaBuildpack);
+
+fn create_scala_build_plan() -> BuildPlan {
+    BuildPlanBuilder::new()
+        .requires("jdk")
+        .provides("jvm-application")
+        .requires("jvm-application")
+        .build()
+}
+
+fn create_coursier_cache_layer(
+    context: &BuildContext<ScalaBuildpack>,
+    env: &Env,
+    build_config: &BuildConfiguration,
+) -> Result<Env, Error<ScalaBuildpackError>> {
+    let coursier_cache_layer = context.handle_layer(
+        layer_name!("coursier_cache"),
+        CoursierCacheLayer {
+            available_at_launch: build_config.sbt_available_at_launch,
+        },
+    )?;
+    Ok(coursier_cache_layer.env.apply(Scope::Build, env))
+}
+
+fn create_ivy_cache_layer(
+    context: &BuildContext<ScalaBuildpack>,
+    env: &Env,
+    build_config: &BuildConfiguration,
+) -> Result<Env, Error<ScalaBuildpackError>> {
+    let ivy_cache_layer = context.handle_layer(
+        layer_name!("ivy_cache"),
+        IvyCacheLayer {
+            available_at_launch: build_config.sbt_available_at_launch,
+        },
+    )?;
+    Ok(ivy_cache_layer.env.apply(Scope::Build, env))
+}
+
+fn create_sbt_layer(
+    context: &BuildContext<ScalaBuildpack>,
+    env: &Env,
+    build_config: &BuildConfiguration,
+) -> Result<Env, Error<ScalaBuildpackError>> {
+    log_header("Installing sbt");
+    let sbt_layer = context.handle_layer(
+        layer_name!("sbt"),
+        SbtLayer {
+            sbt_version: build_config.sbt_version.clone(),
+            sbt_opts: build_config.sbt_opts.clone(),
+            available_at_launch: build_config.sbt_available_at_launch,
+            env: env.clone(),
+        },
+    )?;
+    Ok(sbt_layer.env.apply(Scope::Build, env))
+}
+
+// the native package plugin produces binaries in the target/universal/stage directory which is not included
+// in the list of directories to clean up at the end of the build since a Procfile may reference this
+// location to provide the entry point for an application. wiping the directory before the application build
+// kicks off will ensure that no leftover artifacts are being carried around between builds.
+fn cleanup_any_existing_native_packager_directories(app_dir: &Path) {
+    let native_package_directory = app_dir.join("target").join("universal").join("stage");
+    if native_package_directory.exists() {
+        let delete_operation = create_file_tree(native_package_directory).delete();
+        if delete_operation.is_err() {
+            log_warning(
+                "Removal of native package directory failed",
+                formatdoc! {"
+                    This error should not affect your built application but it may cause the container image
+                    to be larger than expected.
+                "},
```

Comment:
this is an exceptional case.  this is a compilation artifact that is created during the build and is not something that should fail.  and, if it does start failing, it can be traced back to this specific code.